cloud_functions/extract_embedding.py

```diff
@@ -0,0 +1,143 @@
+#
+# Unless required by applicable law or agreed to in writing, softwar3
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This function creates embeddings from a given image input.
+This cloud function first saves all necessary components of the DELG model to t
+mp directories from GCP storage buckets and after doing so uses tensorflow to l
+oad and run the saved model on the image.
+"""
+
+
+import os
+import tempfile
+import file_path_utils as fp
+from google.cloud import storage
+from file_path_utils import get_user_and_dataset_name, get_photo_name
+import cv2
+import numpy as np
+import tensorflow as tf
+
+REQUIRED_SIGNATURE = 'serving_default'
+REQUIRED_OUTPUT = 'global_descriptor'
+EMBEDDING_BUCKET_NAME = 'embeddings_visualizer_output_bucket'
+EMBEDDINGS_FOLDER_NAME = 'embeddings'
+MAX_RESOLUTION = 800
+# SETUP (runs once, when deploying):
+
+# Storage API
+storage_client = storage.Client()
+
+# Create tmp dirs for DELG SavedModel. (Hard-drive is locked, must download all files to /tmp!) 
+os.mkdir('/tmp/saved_model')
+os.mkdir('/tmp/saved_model/variables')
+
+# Download DELG SavedModel files:
+model_bucket = storage_client.bucket('delg_model_bucket')
+saved_model_blob = model_bucket.get_blob('saved_model.pb')
+saved_model_blob.download_to_filename('/tmp/saved_model/saved_model.pb')
+variable_index_blob = model_bucket.get_blob('variables/variables.index')
+variable_index_blob.download_to_filename('/tmp/saved_model/variables/variables.index')
+variable_data_blob = model_bucket.get_blob('variables/variables.data-00000-of-00001')
+variable_data_blob.download_to_filename('/tmp/saved_model/variables/variables.data-00000-of-00001')
+
+# Load DELG from SavedModel dir.
+model = tf.saved_model.load('/tmp/saved_model/')
+
+# Make a fn() from the loaded model.
+embedding_fn = model.signatures[REQUIRED_SIGNATURE]
+
+
+# Entry point: triggered when file is added to the input bucket
+def extract_and_save_embedding(file_data, context):
+    """Extracts and saves embeddings.
+    Runs locally saved delg model on each image, and uploads the image blob
+    to storage.
+    Args:
+        file_data:
+            Dictionary that contains data specific to the event.
+        context:
+            Cloud functions eventmetadata.
+    """
+    image = load_image(file_data['bucket'], file_data['name'])
+    embedding = get_embedding(image)
+    save_embedding(file_data['name'], embedding)
+
+
+def load_image(bucket_name, image_file_name):
+    """Loads the actual image itself.
+    Retrieves the image from storage, and resizes it if neccessary.
+    Args:
+        bucket_name:
+            GCP bucket for the image.
+        image_file_name:
+            Name of the image itself.
+    """
+    tmp_image_filename = os.path.join(
```

Comment:
So looks like this is a dir name, not a file name. Then, shouldn't this be "tmp_image_file_dir"?
But then it is used by "download_to_filename" so I don't really know how it works. Can you explain?

Reply:
maybe it should be called tmp_image_filepath? because it is the directory, but we also add the filename with the os.path.basename() command, and the .download_to_filename() command takes the filepath string (for context, download_to_file takes a file object vs filename taking the filepath string)

Reply:
Oh, it's basename not dirname. My bad. Feel free to keep it or change it to tmp_image_filepath. Sorry about it.